### ISSUE TITLE
[HOLD for payment 2023-05-03] [HOLD for payment 2023-04-26] [$32,000] Allow keyboard shortcuts to work on native devices

### ISSUE THREAD

**Issue by jasperhuangg** [Expensify/App#6883]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Try to use <Command, K> to access the chat switcher on my tablet.

## Expected Result:
The chat switcher should open

The goal is to let native listen and react to hardware keyboard shortcuts similar to how it works in web.

## Actual Result:
Keyboard shortcuts doesn't work. It doesn't open.

## Workaround:
Nope.

## Platform:
Where is this issue occurring?

- iOS
- Android

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @sophiepintoraetz (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by parasharrajat** [comment]
@jasperhuangg Don't forget to revert https://github.com/Expensify/App/pull/6881 which hides the shortcut button from the settings page.

**Comment by MelvinBot** [comment]
Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MelvinBot** [comment]
8 days overdue is a lot. Should this be a Weekly issue? If so, feel free to change it!

**Comment by mananjadhav** [comment]
@jasperhuangg Is there any progress here? If you're fine and like to make it External, I can help with this.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @NicMendonca (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`Exported`)

**Comment by mananjadhav** [comment]
Thanks, @luacmartins. I'll send in a proposal in a few hours.

**Comment by NicMendonca** [comment]
Hi @mananjadhav 👋  ☝️  Are you still committed to working on this?

**Comment by mananjadhav** [comment]
Thanks for the ping. I won't be able to soon.  If somebody doesn't pick in a week, then I'll check again.

I'll share my progress so far if it helps anyone. There are three parts to the implementation.
1. Enabling views/shortcut modal etc in all devices
2. At the moment, the keyboard shortcut map is created after the shortcut is registered, I feel we can create it purely from the `CONST` file.
3. Adding the shortcuts for Native devices with attachable external keyboards will not be done by the current shortcut implementation. It'll need `UIKeycommand` implementation. There are two packages that I found [react-native-keycommands](https://github.com/RocketChat/react-native-keycommands) and  [@envoy/react-native-key-commands](https://github.com/envoy/react-native-key-commands). I haven't tested them.

**Comment by luacmartins** [comment]
Still looking for contributors

**Comment by NicMendonca** [comment]
doubled amount: https://www.upwork.com/jobs/~0190053c928a98fb14

**Comment by luacmartins** [comment]
Still looking for contributors

**Comment by NicMendonca** [comment]
doubled: https://www.upwork.com/jobs/~0190053c928a98fb14

**Comment by luacmartins** [comment]
Still looking for contributors

**Comment by NicMendonca** [comment]
doubled again!

**Comment by luacmartins** [comment]
Still looking for contributors

**Comment by luacmartins** [comment]
Still looking for contributors.

**Comment by NicMendonca** [comment]
Doubled: https://www.upwork.com/jobs/~0190053c928a98fb14

**Comment by NicMendonca** [comment]
doubled - https://www.upwork.com/jobs/~0190053c928a98fb14

**Comment by NicMendonca** [comment]
doubled - https://www.upwork.com/jobs/~0190053c928a98fb14

**Comment by mananjadhav** [comment]
@luacmartins It seems there are no new proposals. Can you review this comment for the approach? https://github.com/Expensify/App/issues/6883#issuecomment-1042130274 If this sounds acceptable as a proposal then I can work on this.

My only concern is I don't know how to test. I don't have an iPad to test, nor we allow contributors to build iOS app.

**Comment by luacmartins** [comment]
@mananjadhav Could you expand more on each step? Why are each of those necessary? What code changes are you proposing?

**Comment by NicMendonca** [comment]
😱 doubled -- https://www.upwork.com/jobs/~0190053c928a98fb14

**Comment by kidroca** [comment]
Can we clarify that a physical/hardware keyboard is needed - there's no "command" button on on-screen keyboards and on-screen keyboards appear only when an input field is focused

The goal here is to let `native` listen and react to hardware keyboard shortcuts similar to how it works in web, and not supporting shortcuts from the on-screen keyboard, correct?

**Comment by mananjadhav** [comment]
## Proposal

As I mentioned in this comment, there are three parts to the implementation (with the 3rd one being most tedious/uncertain).


1. Showing up KeyboardShortcuts modal in all platforms - It looks like this already exists and we don't need to change in the view logic of the KeyboardShortcuts
2. The shortcuts in the KeyboardShortcuts modal are shown from the map built in `libs/KeyboardShortcut/index.js:getDocumentedShortcuts`

This `documentedShortcuts` map is built only when the events are registered/subscribed to the shortcuts. I feel we can build a map purely from `CONST` file.

```
  function getDocumentedShortcuts() {
  	const shortcutsInConfig = _.values(CONST.KEYBOARD_SHORTCUTS);

  	const documentedShortCuts = lodash.map(shortcutsInConfig, shortcut => {
  		const platformAdjustedModifiers = getPlatformEquivalentForKeys(shortcut.modifiers);
  		const displayName = getDisplayName(shortcut.shortcutKey, platformAdjustedModifiers);
  		
  		return {
            shortcutKey: shortcut.shortcutKey,
            descriptionKey: shortcut.descriptionKey ,
            displayName,
            modifiers,
        };
  	});
  		
  }

```

This will also allow us to refactor `subscribe` function to only pass shortcutKey and maybe remove `description` and modifiers.



3. For `KeyboardShortcuts/index.native.js` we add an implementation which subscribes to the shortcuts.

We add the shortcuts using `https://github.com/adamtew/react-native-key-commands` package. I don't have anyway of testing this as of now, and hence I am refering to their documentation for the usage. In the actual implementation I'll fork the above package to add the shortcuts via actions instead of a component

We use the forked version of the `KeyCommand` component to replicate the subscribe and unsubscribe functions.

```
    const eventHandlers = {}


    function subscribe(key, callback, modifiers) {
    	const { KeyCommandEvent } = NativeModules;
    	const eventEmitter = new NativeEventEmitter(KeyCommandEvent);
    	const subscription = eventEmitter.addListener('KeyCommandEvent', callback);
    	NativeModules.KeyCommandsView.registerKeyCommand('KeyCommandEvent', key, getPlatformEquivalentForKeys(modifiers));
    	const platformAdjustedModifiers = getPlatformEquivalentForKeys(shortcut.modifiers);
  		const displayName = getDisplayName(shortcut.shortcutKey, platformAdjustedModifiers);
    	const callbackID = Str.guid();
    	eventHandlers[displayName] ={
    		callbackID: callbackID,
    		subscription,
    	}; 
    	return () => unsubscribe(displayName, callbackID);
    }


    function unsubscribe(displayName, callbackID) {
    	const callback =  _.find(eventHandlers[displayName], callback => callback.id === callbackID);
    	callback.subscription.remove();
 	   	eventHandlers[displayName] = _.reject(eventHandlers[displayName], callback => callback.id === callbackID);
	}


	function getPlatformEquivalentForKeys(modifiers) {
		return {
			'SHIFT': KeyCommandModifiers.ModifierShift,
			....
		}
	}
````

**Comment by azimgd** [comment]
Proposal:

1. Implement a native module using Obj-C and Java that registers and listens to specified keyboard events, and dispatches payload from native to JS.
2. Add support for the native module into [KeyboardShortcut/index.native.js](https://github.com/Expensify/App/blob/b3faacce2749f620fa91a764f6e7ab91bccee0dd/src/libs/KeyboardShortcut/index.native.js)

# Native module
I wrote a sample RN library that allows us to capture hardware (+ simulator) keyboard events and propagates a payload to JS that you can see here: https://github.com/azimgd/keycommandlib.
The motivation behind this was that there are no active and maintained libraries (last commit 2-3 years ago) that work with the recent versions of RN and are multiplatform (Android + iOS). Tested 3 libraries Manan mentioned above, and none worked well.

It mounts on the root view and triggers `onKey` whenever keyboard is pressed. Couple caveats here:
 - `onKey` from the native module should have event bubbling support. Meaning that `onKey` works even when user is focused on TextInput.
 - iOS is implemented using event method swizzling. Get event and propagate further, similar approach is done with Android. 
 
 Module consists of 3 main methods:
 - `registerKeyEvent`: accepts: [input char, modifier flag keycode], registers keyevent. Duplicated execution is handled.
 - `unregisterKeyEvent`: unregisters.
 - `getConstants`: modifier flags (CMD, Ctrl, Alt, ArrowUp ...`) constants that vary per platform.
 - additionally integrate https://developer.apple.com/documentation/uikit/uiresponder/1621141-keycommands?language=objc
 
 ```js
React.useEffect(() => {
    registerKeyCommand([
      {
        input: 'k',
        modifierFlags: constants.keyModifierCommand,
      },
    ]);

    const eventListener = eventEmitter.addListener('onKeyCommand', (event) => {
      // fire an action here
    });

    return eventListener.remove;
}, []);
 ```
 
 
 # JS Side
Add support for the library at: KeyboardShortcut/index.native.js. Shortcuts are already implemented in the web version it should work out of the box after when we add register + subscribe capability from the step above added into index.native file.
- Fetch the list of shortcuts specified at `CONST.js`
- Treat `Ctrl` as `Cmd` on ios
- Register / unsubscribe callback

https://user-images.githubusercontent.com/4882133/167899950-452696f4-037f-4a48-acb7-1dfbed9097ba.mp4


https://user-images.githubusercontent.com/4882133/167902269-03cd2eca-6776-4a4d-bb9b-71ca656942ee.mp4


Tested on both real device and the simulator.
https://github.com/azimgd/keycommandlib/blob/master/ios/Keycommandlib.m
https://github.com/azimgd/keycommandlib/blob/master/android/src/main/java/com/keycommandlib/KeycommandlibModule.java (Android version needs a bit more work)

**Comment by parasharrajat** [comment]
Thanks for the proposal everyone. @mananjadhav I see that https://github.com/adamtew/react-native-key-commands#android is not supported on Android.

@azimgd Great job on the package. Are you planning on maintaining it?  

> onKey from the native module should have event bubbling support. Meaning that onKey works even when user is focused on TextInput.
    iOS is implemented using event method swizzling. Get event and propagate further, similar approach is done with Android.
What are your plans on the caveats? Any roadmaps. I think we can work with capturing instead of bubbling if that works.

**Comment by azimgd** [comment]
Thanks @parasharrajat. I do plan to maintain it. We can also think on:
- Integrate the changes directly into expensify/app
- Transfer repo and get the package released under @expensify

Don't see any caveats, R&D and implementation for the first version showed that the things should be straightforward.

Current roadmap:
- Implement iOS version using current `keycommandlib` version (which works pretty well on iOS).
- Fix last bits on Android version
  - Cache registered keys
  - Better subscribe/unsubscribe logic
- Integrate and Android
- Test both iOS and Android on real hardware keyboard

<details>
  <summary>POC I have been working on:</summary>
  
```js
import {registerKeyCommand, eventEmitter, constants} from 'keycommandlib';
import _ from 'underscore';
import lodashGet from 'lodash/get';
import Str from 'expensify-common/lib/str';
import getOperatingSystem from '../getOperatingSystem';
import CONST from '../../CONST';

// Handlers for the various keyboard listeners we set up
const eventHandlers = {};

// Documentation information for keyboard shortcuts that are displayed in the keyboard shortcuts informational modal
const documentedShortcuts = {};

/**
 * Generates the normalized display name for keyboard shortcuts.
 *
 * @param {String} key
 * @param {String|Array<String>} modifiers
 * @returns {String}
 */
function getDisplayName(key, modifiers) {
    let displayName = [key.toUpperCase()];
    if (_.isString(modifiers)) {
        displayName.unshift(modifiers);
    } else if (_.isArray(modifiers)) {
        displayName = [..._.sortBy(modifiers), ...displayName];
    }

    displayName = _.map(displayName, modifier => lodashGet(CONST.KEYBOARD_SHORTCUT_KEY_DISPLAY_NAME, modifier.toUpperCase(), modifier));

    return displayName.join(' + ');
}

/**
 * Return platform specific modifiers for keys like Control (CMD on macOS)
 *
 * @param {Array<String>} keys
 * @returns {Array}
 */
function getPlatformEquivalentForKeys(keys) {
    const operatingSystem = getOperatingSystem();

    return _.map(keys, (key) => {
        if (!_.has(CONST.PLATFORM_SPECIFIC_KEYS, key)) {
            return key;
        }

        const platformModifiers = CONST.PLATFORM_SPECIFIC_KEYS[key];
        return lodashGet(platformModifiers, operatingSystem, platformModifiers.NATIVE || key);
    });
}

function getKeyEventModifiers(event) {
    const modifiers = [];

    if (event.modifierFlags === constants.keyModifierShift) {
        modifiers.push('SHIFT');
    }
    if (event.modifierFlags === constants.keyModifierCommand) {
        modifiers.push('CONTROL');
    }
    if (event.modifierFlags === constants.keyModifierAlternate) {
        modifiers.push('ALT');
    }
    // if (event.modifierFlags === constants.keyModifierCommand) {
    //     modifiers.push('META');
    // }
    return modifiers;
}

function bindHandlerToKeydownEvent(event) {
    const eventModifiers = getKeyEventModifiers(event);
    const displayName = getDisplayName(event.input, eventModifiers);

    // Loop over all the callbacks
    _.every(eventHandlers[displayName], (callback) => {
        if (_.isFunction(callback.callback)) {
            callback.callback(event);
        }
        return false;
    });
}

eventEmitter.addListener('onKeyCommand', bindHandlerToKeydownEvent);

/**
 * This is a no-op component for native devices because they wouldn't be able to support keyboard shortcuts like
 * a website.
 */
const KeyboardShortcut = {
    subscribe(key, callback, descriptionKey, modifiers = 'shift', captureOnInputs = false, shouldBubble = false, priority = 0, shouldPreventDefault = true) {
        const platformAdjustedModifiers = getPlatformEquivalentForKeys(modifiers);
        const displayName = getDisplayName(key, platformAdjustedModifiers);
        if (!_.has(eventHandlers, displayName)) {
            eventHandlers[displayName] = [];
        }

        const callbackID = Str.guid();
        eventHandlers[displayName].splice(priority, 0, {
            id: callbackID,
            callback,
            captureOnInputs,
            shouldPreventDefault,
            shouldBubble,
        });

        if (descriptionKey) {
            documentedShortcuts[displayName] = {
                shortcutKey: key,
                descriptionKey,
                displayName,
                modifiers,
            };
        }

        registerKeyCommand([
            {
                input: key,
                modifierFlags: constants.keyModifierCommand,
            },
        ]);

        return () => {};
    },
    getDocumentedShortcuts() { return []; },
};

export default KeyboardShortcut;
```
</details>

**Comment by parasharrajat** [comment]
Ok. Thanks. Although @mananjadhav's proposal is quicker and easier, docs say that lib lacks support for iOS. On the other hand, I liked the idea of creating the package. @azimgd's proposal looks good to me. 

cc: @luacmartins 


:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by luacmartins** [comment]
>The goal here is to let native listen and react to hardware keyboard shortcuts similar to how it works in web, and not supporting shortcuts from the on-screen keyboard, correct?

@kidroca correct! I added that to the issue description.

@mananjadhav as Rajat mentioned, that package has no support for Android, so we should keep looking.

**Comment by luacmartins** [comment]
@azimgd thanks for the detailed proposal and @parasharrajat for the review. I'll analyze this further and ask for a few more eyes on it as well.

**Comment by ahmdshrif** [comment]
>  I see that https://github.com/adamtew/react-native-key-commands#android is not supported on Android.

@parasharrajat  
why not use the key-command lib for ios and react-native-key-event for android (it works for android only ). 
https://github.com/Yufeikang/react-native-key-event

if you ok with this approach I can make POC. 

there is also library support both but after the test. it's not working fine for multiple keypresses on ios.  
https://github.com/kevinejohn/react-native-keyevent

**Comment by parasharrajat** [comment]
@ahmdshrif we already have good proposals and I am confident to go either of those. Thanks for your interest.

**Comment by luacmartins** [comment]
@azimgd's proposal looks good! However, I think that we would move forward with the package being released as a `Expensify/*` library, but still remain open source.

Assigning @azimgd to this issue and working on getting the repo created under Expensify.

**Comment by luacmartins** [comment]
Here's the repo https://github.com/Expensify/react-native-key-command

**Comment by azimgd** [comment]
@luacmartins I need a write access to the repo above 👆

**Comment by parasharrajat** [comment]
Maybe fork it and send PRs.

**Comment by luacmartins** [comment]
Yea, I think fork + PRs should work for now. Thanks for the suggestions @parasharrajat! Let me know if you have any issues!

**Comment by azimgd** [comment]
Expecting to push a PR and update by Mon-Tue next week.

**Comment by luacmartins** [comment]
@azimgd any updates?

**Comment by luacmartins** [comment]
Nice! Thanks for the update. Let me know when things are ready to be reviewed!

**Comment by parasharrajat** [comment]
I will review this. 

A few other things:

- Please break your implementation in multiple PRs. You can use any number of PR you want. 
- To reduce the time, create a plan or roadmap on features you are going to implement on the lib. I suggest using the project and issues to track milestones. This way we can also track progress.

Reviewing 85 files at once will be a hell for anyone. 

First create a initial PR with a POC (this one could be large). Then use small PRs to improve it. 

Initially, we can only focus on making the lib suitable for E/app. Later things can be generalized. I am sure that when this would be opened to open source community, it will get more traction and help to make it work for everyone.

**Comment by luacmartins** [comment]
Thanks for jumping in the review and the tips @parasharrajat! I really appreciate it!

**Comment by parasharrajat** [comment]
Review started on the PR. It would be great to get this [info](https://github.com/Expensify/App/issues/6883#issuecomment-1150357942).

**Comment by luacmartins** [comment]
@parasharrajat how's the review going?

**Comment by parasharrajat** [comment]
I am still waiting on https://github.com/Expensify/App/issues/6883#issuecomment-1150357942 details. And some instructions to test the PR. Currently, PR is in a draft state.

**Comment by parasharrajat** [comment]
Thank you @azimgd

**Comment by luacmartins** [comment]
Any updates here?

**Comment by NicMendonca** [comment]
Heads up! I am going OOO so am reassigning

**Comment by azimgd** [comment]
Hey @luacmartins and @parasharrajat. Everything is almost ready from my side. I started with Rajat's suggestion of creating atomic PR's to simplify CR process for you. Here is the first PR https://github.com/Expensify/react-native-key-command/pull/2 with the boilerplate only code. Please go ahead and start reviewing it. I will then move the code from the first PR as soon as it's merged.

**Comment by luacmartins** [comment]
I'll let @parasharrajat review it first!

**Comment by parasharrajat** [comment]
@luacmartins I reviewed the PR, you can also take a look.

**Comment by luacmartins** [comment]
Yes, seems like we have some unaddressed comments.

**Comment by luacmartins** [comment]
First PR in review!

**Comment by kadiealexander** [comment]
Hey team! Any updates?

**Comment by azimgd** [comment]
Hey, blocked / waiting for https://expensify.slack.com/archives/C01GTK53T8Q/p1659024346154599 to be answered

**Comment by luacmartins** [comment]
@azimgd answered in the PR!

**Comment by luacmartins** [comment]
First PR merged!

**Comment by luacmartins** [comment]
We have another [PR](https://github.com/Expensify/react-native-key-command/pull/3) up

**Comment by luacmartins** [comment]
PR above is still in review

**Comment by kadiealexander** [comment]
Just adding a note that I'm OOO until September 21st. As this is on hold I won't reassign it, but if you need a contributor manager to do anything before that point please reapply the `external` label.

**Comment by kadiealexander** [comment]
Waiting for deployment, not overdue!

**Comment by luacmartins** [comment]
Nice! I marked the PR as ready for review!

**Comment by luacmartins** [comment]
3rd PR is in review!

**Comment by luacmartins** [comment]
https://github.com/Expensify/react-native-key-command/pull/4 in review

**Comment by parasharrajat** [comment]
What is next @azimgd ?

**Comment by azimgd** [comment]
Working on a new PR against E/App. @luacmartins could you work towards deploying https://github.com/expensify/react-native-key-command on NPM please ?

**Comment by parasharrajat** [comment]
Let's test this out first with NewDot and then we can prepare to release. May be both can be done parallel.

**Comment by luacmartins** [comment]
@azimgd can we add a wrapper for web so that we have a unified API and don't need to split our files into `index.js` and `index.native.js`?

**Comment by azimgd** [comment]
> Can we add a wrapper for web so that we have a unified API and don't need to split our files into index.js and index.native.js?

JS Implementation of the same functionality you mean ?

**Comment by luacmartins** [comment]
Yea, we should be able to import the module once and it should export the native functionality if it's a native platform or the web functionality if it's web/desktop.

You can see how we did that for Onyx by following the entry points for the lib - https://github.com/Expensify/react-native-onyx/blob/4f2cd795499d3a94ffea78812a461c7baad38f55/package.json#L31-L33

**Comment by parasharrajat** [comment]
Good question. @luacmartins  Do we need to implement the web platform as well for this lib? I understand that you are asking to creating noop web exports to keep it consistent for RN projects. 


@azimgd As we are progressing towards Fabric, we might have to add turbo modules implementation as well.

**Comment by luacmartins** [comment]
>I understand that you are asking to creating noop web exports to keep it consistent for RN projects.

This

**Comment by azimgd** [comment]
What if we port current KeyboardShortcuts implementation within this project, so that it becomes truly cross platform?

**Comment by luacmartins** [comment]
That would work too!

**Comment by azimgd** [comment]
Working on a new scope for this issue per discussion with @luacmartins. New milestone is to create a web version for keycommand package.

Ready to be reviewed:
https://github.com/Expensify/react-native-key-command/pull/8

Work in progress:
https://github.com/Expensify/react-native-key-command/pull/9

**Comment by luacmartins** [comment]
https://github.com/Expensify/react-native-key-command/pull/8 is merged.

**Comment by parasharrajat** [comment]
@azimgd Any update?

**Comment by azimgd** [comment]
https://github.com/Expensify/react-native-key-command/pull/10

**Comment by luacmartins** [comment]
@azimgd is the last PR ready to be reviewed?

**Comment by azimgd** [comment]
Yeap

**Comment by luacmartins** [comment]
Asked for an update [here](https://github.com/Expensify/react-native-key-command/pull/10)

**Comment by luacmartins** [comment]
@azimgd are we ready to start implementing the new lib into App?

**Comment by azimgd** [comment]
Correct, on it now.

**Comment by azimgd** [comment]
@luacmartins not able to sign-in for the last couple days, added my IP address into gdoc earlier yesterday.

**Comment by luacmartins** [comment]
@azimgd thanks for the update! Hopefully we'll get this issue sorted soon!

**Comment by luacmartins** [comment]
@azimgd have you been able to signin?

**Comment by azimgd** [comment]
It got fixed just couple hours ago (had issues with for couple weeks).

**Comment by azimgd** [comment]
It got fixed just couple hours ago (had issues with for couple weeks).

**Comment by luacmartins** [comment]
@azimgd glad it got fixed. Let's push this through the finish line!

**Comment by azimgd** [comment]
Haha, spent a week trying to understand why “cmd + k” (and couple other specific combinations) don’t, while “cmd + f” etc work on simulator with expensify app. Until I re-tested on a hardware device where everything works just fine. 🤷‍♂️ Those are actually reserved keywords which don’t bubble up. 

@luacmartins @parasharrajat I will push my code by the end of this week, so when you test iOS beware of this fact and please test on an iPhone.

**Comment by luacmartins** [comment]
Thanks for the update @azimgd

**Comment by luacmartins** [comment]
@azimgd did you manage to work on this some more?

**Comment by azimgd** [comment]
Integrated both, will create a PR tomorrow.

Should I match the designs to web version? Because right now, shortcuts modal is completely broken on mobile.

<img width="320" alt="Screen Shot 2023-01-10 at 02 48 36" src="https://user-images.githubusercontent.com/4882133/211415704-46f737cb-e991-4b6a-8fad-6846be4002d9.png">

**Comment by luacmartins** [comment]
Yes, please match them to the web version.

**Comment by azimgd** [comment]
@luacmartins could you deploy https://github.com/Expensify/react-native-key-command to npm ?

**Comment by azimgd** [comment]
@luacmartins what's the easiest way for me to test PR above on a real device (iOS) ?

**Comment by luacmartins** [comment]
@azimgd can you change the package name to `@expensify/react-native-key-command` we are trying to standardize on using the `@expensify` scope for our new packages.

**Comment by luacmartins** [comment]
We can build the PR for testing

**Comment by azimgd** [comment]
for sure, let me know once `@expensify/react-native-key-command` is published.

**Comment by luacmartins** [comment]
@azimgd we decided to keep the package name outside the Expensify scope since our other original libs like `react-native-onyx` are also published outside the scope.

That being said, I tried to publish the package today, but I wasn't able to because the package was already [published](https://www.npmjs.com/package/react-native-key-command) about a month ago. Could I ask you to please un-publish the package so we can move forward with publishing it from the Expensify org?

**Comment by azimgd** [comment]
Needs to be reviewed and merged: https://github.com/Expensify/react-native-key-command/pull/12

**Comment by azimgd** [comment]
I'm not able to find a way to have "Enter (Return)" key for ios.
https://developer.apple.com/documentation/uikit/uikeycommand/input_strings_for_special_keys?language=objc
https://developer.apple.com/documentation/uikit/uikeymodifierflags?language=objc

Here is the list of all key-commands we support (iOS):
https://github.com/Expensify/react-native-key-command/blob/main/ios/KeyCommand.mm#L8-L27

@luacmartins what do you suggest ?
@sketchydroide maybe you have an idea ?

**Comment by luacmartins** [comment]
I'll check this tomorrow!

**Comment by sketchydroide** [comment]
@azimgd no sorry, never dealt with anything like this before, key shortcuts are not normally a thing in Native development, and I think it's more now that iPad has become something more like a computer that it starts to make more sense.

**Comment by luacmartins** [comment]
@azimgd I'm not too sure, but maybe [this page](https://www.grapecity.com/blogs/how-to-implement-keyboard-shortcuts-for-flexgrid-ios) could help?

It mentions some keys having dedicated constants, but others such as return/enter use `@"\\r"`

**Comment by azimgd** [comment]
That's for textfield I believe.

**Comment by luacmartins** [comment]
What do you mean by textfield?

**Comment by azimgd** [comment]
That’s not applicable for swizzled keyboard event.
However, I just realized that real device might output a result instead of nil.

**Comment by azimgd** [comment]
Enter key is fixed

**Comment by luacmartins** [comment]
https://github.com/Expensify/react-native-key-command/pull/12 is merged!

**Comment by azimgd** [comment]
Thanks @luacmartins. Do you mind taking a look at https://github.com/Expensify/react-native-key-command/pull/13 as well ?

**Comment by azimgd** [comment]
Can you please publish next version when you have a chance. We are getting very close.

**Comment by luacmartins** [comment]
published v0.5.0!

**Comment by azimgd** [comment]
Hey @parasharrajat @luacmartins. Do you think we can speed up the review and get this issue completed this week ? I'm happy to synchronize with Rajat and help if needed.

**Comment by luacmartins** [comment]
@azimgd how can I help?

**Comment by azimgd** [comment]
@luacmartins, I guess Rajat will be able to test on android and web only. What was your suggestion to test it on iOS ?

**Comment by azimgd** [comment]
@parasharrajat we need both testing and code-review here.

**Comment by azimgd** [comment]
https://github.com/Expensify/App/pull/14168

**Comment by luacmartins** [comment]
https://github.com/Expensify/App/pull/14168 is in review!

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @azimgd, @luacmartins, @parasharrajat, @kadiealexander eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by luacmartins** [comment]
App PR in review!

**Comment by azimgd** [comment]
We have merged and reverted the original PR due to deploy blocker, more context here: https://github.com/Expensify/App/pull/14767#issuecomment-1513431667

I'm in the process of rewriting obj-c side which should be ready this week.

**Comment by MelvinBot** [comment]
📣 @azimgd! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/231263719-a8a3b7d3-36a0-4d7f-8ad4-cebe1f560fa4.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by azimgd** [comment]
@luacmartins @parasharrajat rewrite is happening here: https://github.com/Expensify/react-native-key-command/pull/22

**Comment by luacmartins** [comment]
Thanks for jumping on this so quickly @azimgd!

**Comment by azimgd** [comment]
PR is ready for review.

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.3.1-3 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/14767


If no regressions arise, payment will be issued on 2023-04-26. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.3.5-6 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/17708


If no regressions arise, payment will be issued on 2023-05-03. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by luacmartins** [comment]
BZ checklist is not applicable in this case, since this is a new feature.

**Comment by azimgd** [comment]
@kadiealexander are we good to go?

**Comment by kadiealexander** [comment]
Yes I believe so! I've sent contracts to @azimgd and @parasharrajat at the new agreed price of $40,000 (discussed [here](https://expensify.slack.com/archives/C01SKUP7QR0/p1676057353867719)).

**Comment by kadiealexander** [comment]
Everyone is paid. @luacmartins I think this issue is too old to have a bugzero regression checklist, do you think it needs a regression test or should I close the issue?
